Support build environment where `<format>` is not available

> In our company, std::format is not available in our production environment even though we use C++20. I was thinking about using the proxy library to replace virtual functions to improve the performance of our program.

_Originally posted by @zhiyongww in https://github.com/microsoft/proxy/issues/287#issuecomment-2811649424_

Hi @mingxwa, thanks for the PR, it works for me right now. However, there was a minor issue that I was able to fix on my end.

The error message:
```
fbcode/scripts/zhiyongww/proxy.h:515:3: error: declaring overloaded 'operator()' as 'static' is a C++23 extension [-Werror,-Wc++23-extensions]
  515 |   ___PRO_STATIC_CALL(void, T&& self, proxy<F>& rhs)
      |   ^
fbcode/scripts/zhiyongww/proxy.h:48:49: note: expanded from macro '___PRO_STATIC_CALL'
   48 | #define ___PRO_STATIC_CALL(__R, ...) static __R operator()(__VA_ARGS__)
```

--- a/proxy.h
+++ b/proxy.h
@@ -20,7 +20,9 @@
 
 #if __STDC_HOSTED__
 #include <atomic>
+#if __has_include(<format>)
 #include <format>
+#endif  // __has_include(<format>)
 #endif  // __STDC_HOSTED__
 
 #if __cpp_rtti >= 199711L
@@ -2058,7 +2060,7 @@ struct weak_conversion_dispatch : cast_dispatch_base<false, true> {
 template <class F>
 using weak_conversion_overload = weak_proxy<F>() const noexcept;
 
-#if __STDC_HOSTED__
+#if __STDC_HOSTED__ && __has_include(<format>)
 template <class CharT> struct format_overload_traits;
 template <>
 struct format_overload_traits<char>
@@ -2097,7 +2099,7 @@ struct format_dispatch {
     return impl.format(self, fc);
   }
 };
-#endif  // __STDC_HOSTED__
+#endif  // __STDC_HOSTED__ && __has_include(<format>)
 
 #if __cpp_rtti >= 199711L
 struct proxy_cast_context {
@@ -2231,7 +2233,7 @@ struct wildcard {
 
 namespace skills {
 
-#if __STDC_HOSTED__
+#if __STDC_HOSTED__ && __has_include(<format>)
 template <class FB>
 using format = typename FB::template add_convention<
     details::format_dispatch, details::format_overload_t<char>>;
@@ -2239,7 +2241,7 @@ using format = typename FB::template add_convention<
 template <class FB>
 using wformat = typename FB::template add_convention<
     details::format_dispatch, details::format_overload_t<wchar_t>>;
-#endif  // __STDC_HOSTED__
+#endif  // __STDC_HOSTED__ && __has_include(<format>)
 
 #if __cpp_rtti >= 199711L
 template <class FB>
@@ -2509,7 +2511,7 @@ struct weak_dispatch : D {
 
 }  // namespace pro
 
-#if __STDC_HOSTED__
+#if __STDC_HOSTED__ && __has_include(<format>)
 namespace std {
 
 template <pro::facade F, class CharT>
@@ -2541,7 +2543,7 @@ struct formatter<pro::proxy_indirect_accessor<F>, CharT> {
 };
 
 }  // namespace std
-#endif  // __STDC_HOSTED__
+#endif  // __STDC_HOSTED__ && __has_include(<format>)
 
 #undef ___PRO_THROW
 #undef ___PRO_NO_UNIQUE_ADDRESS_ATTRIBUTE

